### DISCUSSION TITLE
Weird Dev Guide

### DISCUSSION THREAD

**Discussion by erlend-sh** [muni-town/weird#154] [General]
Weird is a _distributed_ (meaning optionally peer-to-peer and federated) Content Management System (CMS), built to increase the agency of internet users.

It can also be described as a social network for collectivized thoughts. Weird coordinates and connects people around mutual ideas and initiatives; **group projects**.

Due to its p2p underpinnings, Weird is also a very capable sync & storage engine for your *web stuff*.

## Product Pillars

Our technical foundations are guided by three interconnected product pillars that together form a cohesive product:

Weird ID 🪪 – Independent social sign-in
Weird CMS 🎨 – Personal web space creator
Weird Net 🌐 – An actually-social network of *shared purpose*

## High level architecture

The Weird app stack is written in Rust and TypeScript.

![Weird architecture diagram](https://raw.githubusercontent.com/muni-town/weird/main/docs/services.png)

Essential building blocks:

### Frontend

- SvelteKit

### Backend

- SvelteKit
- Rauthy (OIDC)
- Leaf Server
  - Rust
  - Iroh
  - Willow
 
#### Key Connectors

Not essential for groking Weird app development, but useful for broad-view context of Weird as a part of an overarching ‘super app’, composed of several keystone APIs/protocols.

- TakingNames (icann domains)
- Polyphony (polyproto(discord))
- Kitsune (ActivityPub)
- rsky (AtProto)
- Conduit (Matrix)
- Stalwart (email)
- gitoxide (git)

## Roadmap

[Link to full]

1. Personal webpage generator ✔️ 
2. Auth v1 (OIDC) ✔️ 
3. Web-stuff imports ✔️ 
4. Projects & Pages
5. Auth v2 (IndieAuth + FedCM)
6. Websites connector (*net of shared purpose*)
7. Open Social integrations v1 (fediverse)
8. Mutual Peers Grid (p2p net)
9. Auth v3 (Nomadic ID)

**Comment by azzamsa** [comment]
We can take inspiration from `rust-analyzer` documentation. It is top-notch.

- https://github.com/rust-lang/rust-analyzer/tree/master/docs
- https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/README.md
- https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/architecture.md
